src/librustdoc/clean/utils.rs

```diff
@@ -624,3 +624,25 @@ where
     *cx.impl_trait_bounds.borrow_mut() = old_bounds;
     r
 }
+
+crate fn find_closest_parent_module(tcx: TyCtxt<'_>, def_id: DefId) -> Option<DefId> {
+    let mut current = def_id;
+    // The immediate parent might not always be a module.
+    // Find the first parent which is.
+    loop {
+        if let Some(parent) = tcx.parent(current) {
+            if tcx.def_kind(parent) == DefKind::Mod {
+                break Some(parent);
+            }
+            current = parent;
+        } else {
+            debug!(
+                "{:?} has no parent (kind={:?}, original was {:?})",
+                current,
+                tcx.def_kind(current),
+                def_id
+            );
+            break None;
+        }
+    }
```

Comment:
I think the debugging probably isn't necessary anymore.
```suggestion
    while let Some(parent) = tcx.parent(current) {
        if tcx.def_kind(parent) == DefKind::Mod {
            return Some(parent);
        }
        current = parent;
    }
    None
```

Reply:
Also, I think this needs the whole if/else chain, not just the loop: https://github.com/rust-lang/rust/blob/bb178237c5539c75e1b85ab78a8ab902b1f333d5/src/librustdoc/passes/collect_intra_doc_links.rs#L770-L776

Reply:
Done.